analysis/src/Hint.ml

```diff
@@ -119,4 +119,60 @@ let inlay ~path ~pos ~maxLength ~debug =
                | Some value ->
                  if String.length label > value then None else Some result
                | None -> Some result)
-             | None -> None)))
\ No newline at end of file
+             | None -> None)))
+
+let codeLens ~path ~debug =
+  let lenses = ref [] in
+  let push loc =
+    let range = Utils.cmtLocToRange loc in
+    lenses := range :: !lenses
+  in
+  (* Code lenses are only emitted for functions right now. So look for value bindings that are functions,
+     and use the loc of the value binding itself so we can look up the full function type for our code lens. *)
+  let value_binding (iterator : Ast_iterator.iterator)
+      (vb : Parsetree.value_binding) =
+    (match vb with
+    | {
+     pvb_pat = {ppat_desc = Ppat_var _; ppat_loc};
+     pvb_expr = {pexp_desc = Pexp_fun _};
+    } ->
+      push ppat_loc
+    | _ -> ());
+    Ast_iterator.default_iterator.value_binding iterator vb
+  in
+  let iterator = {Ast_iterator.default_iterator with value_binding} in
+  (* We only print code lenses in implementation files. This is because they'd be redundant in interface files,
+     where the definition itself will be the same thing as what would've been printed in the code lens. *)
+  (if Filename.check_suffix path ".res" then
+   let parser =
+     Res_driver.parsingEngine.parseImplementation ~forPrinter:false
+   in
+   let {Res_driver.parsetree = structure} = parser ~filename:path in
+   iterator.structure iterator structure |> ignore);
+  !lenses
+  |> List.filter_map (fun (range : Protocol.range) ->
+         match Cmt.fullFromPath ~path with
+         | None -> None
+         | Some full -> (
+           match
+             References.getLocItem ~full
+               ~pos:(range.start.line, range.start.character + 1)
+               ~debug
+           with
+           | Some {locType = Typed (_, typeExpr, _)} ->
```

Comment:
We then use a helper to get the full type via the compiler artifact for the target file, using the positions where we found variables being assigned functions. This will lead us to the full, correct type definitions for all functions assigned to variables.

A type at a location can be a bunch of different things - we target `Typed` specifically here, because we know that's the only thing our function definition can be. In `Typed`, `typeExpr` is what we're interested in. `typeExpr` comes from the type checker, and we have helpers that can print those type definitions as strings.